File: src/train.py
import numpy as np
import argparse
import pickle as pkl
import os
from multiprocessing import Pool
import multiprocessing as multi
from tqdm import tqdm
import datetime
import time
from scipy import interpolate

class DiscreteDist:
    def __init__(self, p):
        size = p.shape[0] * 10
        cum = np.cumsum(p)
        cum *= size / cum[-1]
        cum = np.r_[0, cum] # p.shape[0] + 1
        n = np.arange(p.shape[0] + 1)
        f = interpolate.interp1d(cum, n)
        dist = f(np.arange(size)).astype("int64")
        dist[dist < 0] = 0
        dist[dist >= size] = size - 1
        self.dist = dist

    def gen(self, size):
        return self.dist[(np.random.rand(size) * self.dist.shape[0]).astype("int64")]

def train(whole_corpus, dim, epoch_num, window_size, word_samp_th, neg_samp_num, lr):
    print(datetime.datetime.now())
    print("The length of corpus: {}".format(whole_corpus.shape[0]))

    print("Counting number of occurrences of the words. ")
    vocab, count = np.unique(whole_corpus, return_counts = True)
    vocab_num = vocab.shape[0]
    print(datetime.datetime.now())
    print("The size of vocabulary: {}".format(vocab_num))

    print("Discarding too rare words.")
    count[count < min_count] = 0
    word_count_seq = count[whole_corpus]
    mask = word_count_seq > 0
    whole_corpus = whole_corpus[mask] # 低頻度語は除去
    word_count_seq = word_count_seq[mask]

    whole_corpus_len = whole_corpus.shape[0]

    starting_lr = lr

    print(datetime.datetime.now())
    print("Initializing. ")
    # ベクトル の初期化
    v1 = (np.random.rand(vocab_num, dim) - 0.5) / dim
    v2 = np.zeros((vocab_num, dim))

    # negative sampling 分布
    distorted_tf = (count ** 0.75)
    neg_dist = DiscreteDist(distorted_tf)

    trained_word_count = 0
    for epoch in range(epoch_num):
        print(datetime.datetime.now())
        print("Stating epoch {} / {}.".format(epoch + 1, epoch_num))
        print("Discarding high-frequency words.")
        prob_discard = 1 - (np.sqrt(word_count_seq / (word_samp_th * whole_corpus_len)) + 1) * (word_samp_th * whole_corpus_len) / word_count_seq
        corpus = whole_corpus[np.random.rand(whole_corpus_len) < prob_discard]
        corpus_len = corpus.shape[0]

        print(datetime.datetime.now())
        print("Training. ")
        for t in tqdm(range(corpus_len)):
            left = t - window_size
            right = t + window_size
            if left < 0 or right >= corpus_len - 1:
                continue
            word = corpus[t]
            context = np.r_[corpus[left : t], corpus[t + 1: right + 1]]
            neg_rv = neg_dist.gen(context.shape[0] * neg_samp_num)
            v1_w = v1[word] # dim
            v2_c = v2[context, :] # context_num, dim
            v2_neg = v2[neg_rv, :].reshape(-1, neg_samp_num, dim) # context_num, neg_samp_num, dim

            w_c = np.dot(v2_c, v1_w) # context_num
            w_neg = np.dot(v2_neg, v1_w) # context_num, neg_samp_num

            sigma = lambda x: 1 / (1 + np.exp(-x))
            # context_num, dim
            d_w = sigma(-w_c)[:, np.newaxis] * v2_c - np.sum(sigma(w_neg)[:, :, np.newaxis] * v2_neg, axis = 1)
            # context_num, dim
            d_c = sigma(-w_c)[:, np.newaxis] * v1_w[np.newaxis, :]
            # context_num, neg_samp_num, dim
            d_neg = -sigma(w_neg)[:, :, np.newaxis] * v1_w[np.newaxis, np.newaxis, :]
            v1[word] += lr * np.sum(d_w, axis = 0)
            v2[context, :] += lr * d_c
            v2[neg.reshape(-1), :] += lr * d_neg.reshape(-1, dim)

            trained_word_count += 1
            lr = max(starting_lr * (1 - trained_word_count / (epoch_num * whole_corpus_len)), starting_lr * 1e-3)

    v1[count == 0, : ] = 0
    return v1

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-c", "--corpus", help="Path to corpus data. ")
    parser.add_argument("-o", "--output", help="Path to file to save the resulting word vectors. ")

    args = parser.parse_args()

    min_count = 5
    window_size = 5
    word_samp_th = 1e-4
    dim = 100
    neg_samp_num = 5
    lr = 0.025

    epoch_num = 5

    print("Loading corpus. ")
    corpus = np.fromfile(args.corpus, dtype="int32")

    corpus = corpus[:22435478] # 1 / 100

    vec = train(corpus, dim, epoch_num, window_size, word_samp_th, neg_samp_num, lr)
    print("Saving word vector. ")
    vec.astype("float64").tofile(args.output)

File: src/train.cpp
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <math.h>
#include <sys/stat.h>
#include <chrono>
#include <thread>
#include <mutex>
#include <unistd.h>
#include <string.h>
#include "xor128.hpp"
#include "conf.hpp"
#include "progress_display.hpp"

const int64_t NEG_DIST_TABLE_SIZE = MAX_VOCAB_NUM * 100;

int32_t whole_corpus[MAX_CORPUS_LEN];
int64_t whole_corpus_len;
int32_t corpus[MAX_CORPUS_LEN];

int32_t count[MAX_VOCAB_NUM];
int32_t vocab_num;

int32_t neg_dist_table[NEG_DIST_TABLE_SIZE];

double v1[MAX_VOCAB_NUM][N_DIM];
double v2[MAX_VOCAB_NUM][N_DIM];

int n_thread;
int epoch_num;
int min_count;
double word_samp_thre;
double learning_rate;
int window_size;
int neg_samp_num;

double sigmoid(double x) noexcept {
    const double EXP_MAX = 10;
    if (x > EXP_MAX) return 1;
    if (x < -EXP_MAX) return 0;
    double expx = exp(x);
    return expx / (1 + expx);
}

int train() {
    printf("Initializing model parameters ... \n");

    xor128 rand_gen;
    for (int w = 0; w < vocab_num; w++) {
        for (int i = 0; i < N_DIM; i++) {
            v1[w][i] = (rand_gen() - 0.5) / N_DIM;
            v2[w][i] = 0;
        }
    }

    printf("Creating %d threads ...\n", n_thread);

    std::thread* threads[MAX_N_THREAD];
    int64_t corpus_len_per_thread = whole_corpus_len / n_thread;

    progress_display prg_disp;

    for (int k = 0; k < n_thread; k++) {
        threads[k] = new std::thread(
        [ &prg_disp ] (int64_t loc_begin, int64_t loc_len, int thread_id) {
            double dw[N_DIM];

            double starting_lr = learning_rate;
            double lr = starting_lr;

            int64_t trained_word_count = 0;
            int32_t* loc_corpus = corpus + loc_begin;

            xor128 rand_gen;

            for (int epoch = 0; epoch < epoch_num; epoch++) {
                int64_t corpus_len = 0;
                int64_t corpus_len_wo_disc = 0;
                for (int64_t t = loc_begin; t < loc_begin + loc_len; t++) {
                    int32_t word = whole_corpus[t];
                    if (count[word] == 0) continue;
                    corpus_len_wo_disc++;

                    double prob = 1 - (sqrt(count[word] / (word_samp_thre * whole_corpus_len)) + 1) * (word_samp_thre * whole_corpus_len) / count[word];
                    if (rand_gen() < prob) continue;

                    loc_corpus[corpus_len] = word;
                    corpus_len++;
                }
                for (int64_t t = 0; t < corpus_len; t++) {
                    const int progress_interval = 2e4;
                    if (t % progress_interval == 0) {
                        double prg = (epoch + 1. * t / corpus_len) / epoch_num;
                        prg_disp.report(trained_word_count, (int64_t)trained_word_count / (prg + 1e-8), thread_id);
                    }

                    int64_t left = t - window_size;
                    int64_t right = t + window_size;
                    if (left < 0 || right >= corpus_len - 1) continue;

                    int32_t word = loc_corpus[t];
                    for (int64_t tc = left; tc <= right; tc++) {
                        if (tc == t) continue;
                        int32_t context = loc_corpus[tc];

                        double w_c = 0; for (int i = 0; i < N_DIM; i++) w_c += v1[word][i] * v2[context][i];
                        double w_c_sig = sigmoid(-w_c);

                        for (int i = 0; i < N_DIM; i++) dw[i] = w_c_sig * v2[context][i];
                        for (int i = 0; i < N_DIM; i++) v2[context][i] += lr * w_c_sig * v1[word][i];

                        for (int l = 0; l < neg_samp_num; l++) {
                            int32_t neg = word;
                            while (neg == word) neg = neg_dist_table[(int64_t)(rand_gen() * NEG_DIST_TABLE_SIZE)];
                            double w_neg = 0; for (int i = 0; i < N_DIM; i++) w_neg += v1[word][i] * v2[neg][i];
                            double w_neg_sig = sigmoid(w_neg);
                            for (int i = 0; i < N_DIM; i++) dw[i] -= w_neg_sig * v2[neg][i];
                            for (int i = 0; i < N_DIM; i++) v2[neg][i] -= lr * w_neg_sig * v1[word][i];
                        }
                        for (int i = 0; i < N_DIM; i++) v1[word][i] += lr * dw[i];
                    }

                    trained_word_count++;
                    lr = std::max(starting_lr * (1 - (double)trained_word_count / (epoch_num * corpus_len_wo_disc)), 1e-4);
                }
            }
            return 0;
        }
        , corpus_len_per_thread * k, corpus_len_per_thread, k);
    }

    printf("Threads started. \n");
    for (int k = 0; k < n_thread; k++) threads[k]->join();
    printf("Training is finished. \n");
    for (int k = 0; k < n_thread; k++) delete threads[k];

    return 0;
}

int count_words() {
    std::fill(count, count + MAX_VOCAB_NUM, 0);
    vocab_num = 0;
    for (int64_t t = 0; t < whole_corpus_len; t++) {
        int32_t word = whole_corpus[t];
        if (word >= MAX_VOCAB_NUM) exit(EXIT_FAILURE);
        count[word]++;
        vocab_num = std::max(vocab_num, word + 1);
    }
    return 0;
}

int discard_infrequent_words() {
    for (int64_t w = 0; w < vocab_num; w++)
        if (count[w] < min_count)
            count[w] = 0;
    return 0;
}

int nullify_infrequent_word_vectors() {
    for (int64_t w = 0; w < vocab_num; w++)
        if (count[w] == 0)
            for (int i = 0; i < N_DIM; i++)
                v1[w][i] = 0;
    return 0;
}

int init_neg_dist_table() {
    double power = 0.75;
    double Z = 0;
    for (int64_t w = 0; w < vocab_num; w++) Z += pow(count[w], power);
    double cum = 0;
    int64_t w = -1;
    for (int64_t i = 0; i < NEG_DIST_TABLE_SIZE; i++) {
        double p = (double)i / NEG_DIST_TABLE_SIZE;
        while (1) {
            if (p < cum / Z) break;
            if (w >= vocab_num - 1) break;
            w++;
            cum += pow(count[w], power);
        }
        neg_dist_table[i] = w;
    }
    return 0;
}

int64_t get_file_size(const char* path) {
    struct stat st;
    if (stat(path, &st)) return -1;
    return st.st_size;
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        printf("%s [ corpus ] [ output ]\n", argv[0]);
        return 0;
    }

    char *in_path = argv[1];
    char *out_path = argv[2];

    min_count = 30;
    epoch_num = 5;
    word_samp_thre = 1e-5;
    learning_rate = 0.025;
    window_size = 5;
    neg_samp_num = 5;
    n_thread = 16;

    int opt;
    while ((opt = getopt(argc, argv, "e:s:l:w:n:t:m:")) != -1) {
        switch(opt) {
            case 'e': {
                char* end;
                epoch_num = strtol(optarg, &end, 10);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 's': {
                char* end;
                word_samp_thre = strtod(optarg, &end);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 'l': {
                char* end;
                learning_rate = strtod(optarg, &end);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 'w': {
                char* end;
                window_size = strtol(optarg, &end, 10);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 'n': {
                char* end;
                neg_samp_num = strtol(optarg, &end, 10);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 't': {
                char* end;
                n_thread = strtol(optarg, &end, 10);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
            case 'm': {
                char* end;
                min_count = strtol(optarg, &end, 10);
                if (strlen(end) > 0) exit(EXIT_FAILURE);
                break;
            }
        }
    }

    if (n_thread > MAX_N_THREAD) exit(EXIT_FAILURE);

    FILE* fp = fopen(in_path, "rb");
    if (fp == NULL) exit(EXIT_FAILURE);

    int64_t file_size = get_file_size(in_path);
    whole_corpus_len = file_size / sizeof(int32_t);
    printf("The length of corpus: %lld\n", (long long)whole_corpus_len);
    if (whole_corpus_len > MAX_CORPUS_LEN) exit(EXIT_FAILURE);

    printf("Loading corpus ...\n");
    fread(whole_corpus, sizeof(int32_t), whole_corpus_len, fp);

    fclose(fp);

    printf("Counting number of occurrences of the words ... \n");
    count_words();
    printf("The size of vocabulary: %d\n", vocab_num);

    printf("Discarding infrequent words ... \n");
    discard_infrequent_words();

    printf("Making negative sampling distribution table ... \n");
    init_neg_dist_table();

    train();

    printf("Saving word vectors ...\n");

    nullify_infrequent_word_vectors();

    fp = fopen(out_path, "wb");
    if (fp == NULL) exit(EXIT_FAILURE);
    fwrite(v1[0], sizeof(double), vocab_num * N_DIM, fp);
    fclose(fp);

    return 0;
}
